lib/utils/search.js

```diff
@@ -0,0 +1,55 @@
+// find a searchTerm, highlighting it (and scrolling to it if currently off screen)
+export default function find(terminal, searchTerm, options = {}) {
+  const dir = options.dir === -1 ? -1 : 1;
+  const sel = terminal.document_.defaultView.getSelection();
+
+  // get focus row
+  let focusRowNode = sel.focusNode;
+  while (focusRowNode && !('rowIndex' in focusRowNode)) {
+    focusRowNode = focusRowNode.parentNode;
+  }
+
+  // get anchor row
+  let anchorRowNode = sel.anchorNode;
+  while (anchorRowNode && !('rowIndex' in anchorRowNode)) {
+    anchorRowNode = anchorRowNode.parentNode;
+  }
+
+  const startRowNode = dir === 1 ? focusRowNode : anchorRowNode;
+
+  const topRowIndex = terminal.scrollPort_.getTopRowIndex();
+  const bottomRowIndex = terminal.scrollPort_.getBottomRowIndex(topRowIndex);
+  const startRow = startRowNode ?
+    startRowNode.rowIndex :
+    (dir === 1 ? topRowIndex : bottomRowIndex);
+  const selNodeOffset = (
+    startRowNode &&
+    startRowNode.textContent.indexOf((dir === 1 ? sel.focusNode : sel.anchorNode).textContent)
+  ) || 0;
+  let curRowOffset = selNodeOffset + (dir === 1 ? sel.focusOffset : sel.anchorOffset - 1);
+  const rowCount = terminal.getRowCount();
+
+  let foundIndex;
+  let curRow;
+  for (let i = 0; i <= rowCount; i++) {
+    // increment/decrement curRow while ensuring it's always 0 <= curRow < rowCount
+    curRow = (startRow + (rowCount * 2) + (dir * i)) % rowCount;
+    const rowText = terminal.getRowText(curRow);
+    foundIndex = rowText[dir === 1 ? 'indexOf' : 'lastIndexOf'](searchTerm, curRowOffset);
+    if (foundIndex !== -1 && foundIndex !== curRowOffset) {
+      break;
+    }
+    curRowOffset = undefined;
+  }
+
+  if (foundIndex !== -1) {
+    terminal.scrollPort_.scrollTo(curRow, 5);
+    terminal.scrollPort_.scheduleRedraw();
+
+    // Make sure the redraw has happened before attempting to highlight the result
+    setTimeout(() => terminal.selectRow(curRow, foundIndex, foundIndex + searchTerm.length), 0);
+
+    return true;
+  }
+  return false;
```

Comment:
The basic find function should probably just return the detailed results of the search.  That will allow it to be composed into other useful functions.  Leave the scrolling and highlighting to another function.